Add a "Middle-Level" API based on Cypher

ilumsden opened the pull request:
Resolves #363 

This PR adds a new API "level" to the Hatchet Query Langauge. This API uses a slightly modified subset of the [Cypher query language](https://www.opencypher.org/).

As with a normal Cypher query, you can think of a middle-level API query as consisting of two parts:
1. Path Specification
2. Condition Specification

## Path Specification

In the path specification, the user specifies the wildcards and order of abstract nodes. This part of the query must start with the "MATCH" keyword. The rest of the specification consists of a series of abstract nodes and edges in the following general form:
```diff
- MATCH (...)-[...]->(...)-[...]->(...)
+ MATCH (...)->(...)->(...)
```
_In the above example, the text in red is the old version of the path specification, and the text in green is the new version._

In the above example, every instance of "..." represents a possible abstract node. However, only the "..." between parentheses must be replaced with actual content. The "..." between square braces can be left empty if desired.

The path specification must contain at least one "node" (content in parentheses) and zero or more "edges" (the `-[]->` syntax). The path specification must also end with a "node".

Each placeholder above (i.e., "...") must contain one of the following:
1. Only a wildcard (same syntax for wildcards as in the high- and low-level APIs) (e.g., `"."`, `"*"`, `"+"`, `1`, `4`, etc)
2. A "variable name" (to be used in the condition section) (e.g., `p`)
3. Both a wildcard and a variable name, separated by a comma (e.g., `"*", p`)

## Condition Specification

The condition specification section is used to specify all the conditions that define what real nodes can be matched to abstract nodes in the Path Specification section. This is similar to the filters of the high- and low-level APIs, but, in the middle-level API, all the conditions are specified in a single statement using boolean operators (e.g., and, or, etc.). Internally, the `CypherQuery` class will convert all these conditions into the individual filters used in the high- and low-level APIs.

The condition specification section starts with the "WHERE" keyword followed by the conditions. Each individual condition has the following form:
```
<VAR NAME>."<ATTRIBUTE NAME>" <OPERATOR> [VAL]
```
In the above specification, content in angle braces (`<>`) represents required user-provided content, and content in square braces represents optional content. This notation is consistent with Cypher with one major exception: the `ATTRIBUTE NAME` field **must** be in quotes. This is due to the fact that node attributes in Hatchet (e.g., "time", "time (inc)", etc.) are not guaranteed to follow variable naming requirements, which is guaranteed in Cypher.

There are two sets of "operators" in the middle-level API: string operators and number operators. The tables below describes the available operators for each type.
**Type-Less Operators:**
| Operator                | Example                    | Description                                                             |
| :----                        | :----                             | :----                                                                         |
| Is None | `p."time" IS NONE` | Checks if attribute value is None |
| Is Not None | `p."time" IS NOT NONE` | Checks if attribute value is not None |

**String Operators:**
| Operator                | Example                    | Description                                                             |
| :----                        | :----                             | :----                                                                         |
| Equivalence          | `p."name" = "main"`    | Checks if attribute value matches right-hand side |
| Starts With            | `p."name" STARTS WITH "MPI"` | Checks if attribute value starts with right-hand side |
| Ends With             | `p."name" ENDS WITH "impl"`     | Checks if attribute value ends with right-hand side  |
| Contains               | `p."name" CONTAINS "RAJA"`    | Checks if attribute value contains right-hand side as a substring |
| Regex Match        | `p."name" =~ "<([\w]+).*>(.*?)<\/\1>"` | Checks if attribute value satisfies the regex expression |

**Number Operators:**
| Operator                | Example                    | Description                                                             |
| :----                        | :----                             | :----                                                                         |
| Equivalence          | `p."time" = 5`              | Checks if attribute value is equal to right-hand side |
| Less-Than             | `p."time (inc)" < 20`   | Checks if attribute value is less than right-hand side |
| Greater-Than        | `p."time (inc)" > 60`   | Checks if attribute value is greater than right-hand side |
| Less-Than or Equal | `p."time (inc)" <= 20`   | Checks if attribute value is less than or equal to right-hand side |
| Greater-Than or Equal | `p."time (inc)" >= 60`   | Checks if attribute value is greater than or equal to right-hand side |
| Is NaN | `p."time" IS NAN` | Checks if attribute value is NaN |
| Is Not NaN | `p."time" IS NOT NAN` | Checks if attribute value is not NaN |
| Is Infinity | `p."time" IS INF` | Checks if attribute value is Infinity |
| Is Not Infinity | `p."time" IS NOT INF` | Checks if attribute value is not Infinity |

**"Other" Operators:**
| Operator                | Example                    | Description                                                             |
| :----                        | :----                             | :----                                                                         |
| Is None                  | `p."attr" IS NONE`        | Checks if attribute value is `None`                          |
| Is Not None                  | `p."attr" IS NOT NONE`        | Checks if attribute value is not `None`                          |

To combine individual conditions, use the three provided binary and unary operators:
1. `AND`
2. `OR`
3. `NOT`

**NOTE: currently, all syntax is _case sensitive_! In the future, I'll probably be modifying it so the keywords (e.g., "MATCH", "WHERE", "ENDS WITH", etc.) are not case sensitive, but for now, make sure keywords are in all caps!**

## Putting Everything Together

Both parts of the middle-level API query should be combined into a string (I recommend a multi-line string). Once this string is created, the user can use it just like a high-level query. So, to filter a `GraphFrame` using a middle-level API query, simply pass this string to the `GraphFrame.filter` function. For example:
```python
path = """MATCH (p)-[]->(2, q)-["*", r]->(s)
WHERE p."time (inc)" >= 30.0 AND NOT q."name" STARTS WITH "b"
AND r."name" =~ "[^b][a-z]+" AND s."name" STARTS WITH "gr"
"""
filtered_gf = gf.filter(path)
```

Internally, the `GraphFrame.filter` function will pass the middle-level API query to the new `CypherQuery` class, which uses the textX library to parse the input and construct a Hatchet query from it.

ilumsden commented:
To keep things consistent going forward, this PR is dependent on #333. So, #333 needs to be merged before this one.

Also, I just noticed that I have to update the `NaryQuery` class from #333 to support the middle-level API queries.

ilumsden commented:
`NaryQuery` stuff is fixed.

ilumsden commented:
Travis for Python 3.5+ is failing because of an issue while building Timemory. This is also being encountered in #272, even though, unlike #272, this PR makes no changes to the `.travis.yml` file."

ilumsden commented:
I just discovered that, if you're using Python 2.7, all middle-level queries **must** be unicode strings. This requirement is introduced by textX. Additionally, we can't use `from __future__ import unicode_literals` because that causes issues with the object oriented design of the compound queries feature. This can be relaxed once Python 2.7 support is dropped, but until then, all middle-level queries must be written as `u"..."` or `u"""..."""`.

ilumsden commented:
Besides the Timemory issue, this PR should be good with respect to Travis. I'll rebase it after #333 is merged.

cscully-allison commented on hatchet/query.py:
```diff
@@ -3,6 +3,8 @@
 #
 # SPDX-License-Identifier: MIT
 
+# from __future__ import unicode_literals, print_function
```
We should delete this commented code.

cscully-allison replied:
We should delete this commented out code.

cscully-allison reviewed (commented):
Overall looks good. 

The one change I suggest is that we add tests for OR and NOT functionality. Every test I saw only used "AND". If I missed an instance of a test like this then let me know.

ilumsden commented:
@cscully-allison I've removed the commented-out code you pointed out. For the testing, there actually already was a test for NOT, but I've added one for OR.

cscully-allison reviewed (approved):
Looks all good! Thanks Ian :)
